In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", 100)
plt.style.use("seaborn-v0_8")


In [3]:
DATA_PATH = "../data/raw/"
FILE_NAME = "default of credit card clients.xls"

df = pd.read_excel(DATA_PATH + FILE_NAME, header=1)

print("Dataset shape:", df.shape)
df.head()


Dataset shape: (30000, 25)


ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1      20000    2          2         1   24      2      2     -1     -1   
1   2     120000    2          2         2   26     -1      2      0      0   
2   3      90000    2          2         2   34      0      0      0      0   
3   4      50000    2          2         1   37      0      0      0      0   
4   5      50000    1          2         1   57     -1      0     -1      0   

   PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  \
0     -2     -2       3913       3102        689          0          0   
1      0      2       2682       1725       2682       3272       3455   
2      0      0      29239      14027      13559      14331      14948   
3      0      0      46990      48233      49291      28314      28959   
4      0      0       8617       5670      35835      20940      19146   

   BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0          0         0       689         0         0         0         0   
1       3261         0      1000      1000      1000         0      2000   
2      15549      1518      1500      1000      1000      1000      5000   
3      29547      2000      2019      1200      1100      1069      1000   
4      19131      2000     36681     10000      9000       689       679   

   default payment next month  
0                           1  
1                           1  
2                           0  
3                           0  
4                           0

In [4]:
df.columns = (
    df.columns
    .str.strip()
    .str.upper()
    .str.replace(" ", "_")
)
df.head()


ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1      20000    2          2         1   24      2      2     -1     -1   
1   2     120000    2          2         2   26     -1      2      0      0   
2   3      90000    2          2         2   34      0      0      0      0   
3   4      50000    2          2         1   37      0      0      0      0   
4   5      50000    1          2         1   57     -1      0     -1      0   

   PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  \
0     -2     -2       3913       3102        689          0          0   
1      0      2       2682       1725       2682       3272       3455   
2      0      0      29239      14027      13559      14331      14948   
3      0      0      46990      48233      49291      28314      28959   
4      0      0       8617       5670      35835      20940      19146   

   BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0          0         0       689         0         0         0         0   
1       3261         0      1000      1000      1000         0      2000   
2      15549      1518      1500      1000      1000      1000      5000   
3      29547      2000      2019      1200      1100      1069      1000   
4      19131      2000     36681     10000      9000       689       679   

   DEFAULT_PAYMENT_NEXT_MONTH  
0                           1  
1                           1  
2                           0  
3                           0  
4                           0

In [5]:
df = df.rename(columns={"DEFAULT_PAYMENT_NEXT_MONTH": "DEFAULT"})
df["DEFAULT"].value_counts()


DEFAULT
0    23364
1     6636
Name: count, dtype: int64

In [6]:
df.describe().T.head(10)


count           mean            std      min       25%       50%  \
ID         30000.0   15000.500000    8660.398374      1.0   7500.75   15000.5   
LIMIT_BAL  30000.0  167484.322667  129747.661567  10000.0  50000.00  140000.0   
SEX        30000.0       1.603733       0.489129      1.0      1.00       2.0   
EDUCATION  30000.0       1.853133       0.790349      0.0      1.00       2.0   
MARRIAGE   30000.0       1.551867       0.521970      0.0      1.00       2.0   
AGE        30000.0      35.485500       9.217904     21.0     28.00      34.0   
PAY_0      30000.0      -0.016700       1.123802     -2.0     -1.00       0.0   
PAY_2      30000.0      -0.133767       1.197186     -2.0     -1.00       0.0   
PAY_3      30000.0      -0.166200       1.196868     -2.0     -1.00       0.0   
PAY_4      30000.0      -0.220667       1.169139     -2.0     -1.00       0.0   

                 75%        max  
ID          22500.25    30000.0  
LIMIT_BAL  240000.00  1000000.0  
SEX             2.00        2.0  
EDUCATION       2.00        6.0  
MARRIAGE        2.00        3.0  
AGE            41.00       79.0  
PAY_0           0.00        8.0  
PAY_2           0.00        8.0  
PAY_3           0.00        8.0  
PAY_4           0.00        8.0

In [7]:
default_rate = df["DEFAULT"].mean()
print(f"Default rate: {default_rate:.2%}")


Default rate: 22.12%


## Final EDA Summary

- Dataset contains ~30,000 credit card clients
- Default rate is ~22%, indicating moderate class imbalance
- No missing values present
- Mostly numerical features
- Suitable for classical ML models and automated comparison